### ISSUE TITLE
Ignore bubbled events that arise during rendering

### ISSUE THREAD

**Issue by garyb** [slamdata/purescript-halogen#472]
I'm not even sure this is possible in a sensible way, but @safareli brought it up as an idea, and it seems like a good one in theory.

**Comment by garyb** [comment]
Motivation...

Going from this:
```
HH.div_
  [ HH.button
      [ HE.onClick (HE.input ToggleState) ]
      [ HH.text "Close"
      ]
  , HH.text "Expanded"
  ]
```

To:
```
HH.div
  [ HE.onClick $ (HE.input ToggleState) ]
  [ HH.text "Open" ]
```

During re-render will cause the `onClick` to fire on the second case as well as the first, as the event handler is patched into the parent during the state change, so the same `onClick` that causes the initial transition is caught twice.

**Comment by safareli** [comment]
In react this works:
```js
import React from 'react';
import { render } from 'react-dom';


class Compact extends React.Component {
  constructor(props) {
    super(props);
    this.state = { isCompact: true }
  }
  render() {
    if (this.state.isCompact) {
      return this.props.compactView({ expand: (e) => {
        console.log("expand", e)
        this.setState({ isCompact: false})
      }})
    } else {
      return this.props.expandView({ compact: (e) => {
        console.log("compact", e)
        this.setState({ isCompact: true })
      }})
    }
    
  }
}

render(
  <Compact
    expandView={({ compact }) => (
      <div>
        <div onClick={compact}>
          compact
        </div>
      </div>
    )}
    compactView={({ expand }) => (
      <div onClick={expand}>
        expand
      </div>
    )}
  />,
  document.getElementById('root')
);
```

https://codesandbox.io/s/nr6ll460wl
https://nr6ll460wl.codesandbox.io/

**Comment by natefaubion** [comment]
React also implements it's own event dispatch system and scheduler 😛

**Comment by safareli** [comment]
I think it's more `Bug` than `Enhancement` :p

**Comment by natefaubion** [comment]
Why is it a bug? It can be annoying in some cases, but it's just how the DOM even system works. The ways around it would be to use pervasive async rendering (which we aren't going to do), or implement our own event dispatch system (which we are unlikely to do).

**Comment by safareli** [comment]
> but it's just how the DOM even system works

Yes, but if one writes code which interacts with DOM directly they will not reuse DOM nodes, and will not have this issue.

I see reusing of dom node as internal performance optimisation trick of a Virtual DOM lib. If nodes are not reused (and new once are created) than this issue will not be relevant.  By using some Virtual DOM lib you don't want to think on that kinda internal details. So way the lib does optimisation is leaked to client of the lib, which makes me think that it's a bug.

---

I think we should make open issue in https://github.com/Matt-Esch/virtual-dom too right @garyb ?

**Comment by natefaubion** [comment]
> Yes, but if one writes code which interacts with DOM directly they will not reuse DOM nodes, and will not have this issue.

This isn't true. You can get the same behavior using jquery by doing something with a parent node synchronously in an event handler.

> I think we should make open issue in https://github.com/Matt-Esch/virtual-dom too right @garyb?

It's not a problem with `virtual-dom` specifically, or any virtual DOM algorithm directly. It's the interaction of synchronous rendering in an event callback, and using the normal DOM event system. To get around it, you would have to do event delagation at the component level, at the least, which amounts to re-implementing event dispatch.
